### ISSUE TITLE
Vector Constructor Size

### ISSUE THREAD

**Issue by mattbru2** [swig/swig#1012]
I am using SWIG for Java/C++ interaction.

In my swig.i file, I have the following:

`%template(BoolVector) std::vector<bool>;`

On the Java side, if I construct a BoolVector with an initial size, like:

`BoolVector myVector = new BoolVector(10);`

I get very odd behavior on the C++ side trying to read that structure. Often, the vector will end up with a different size than what was initially constructed and populated, let's say 9 or 11 elements, instead of 10.

Is this a bug in SWIG or in how I am using SWIG? I am using version 3.0.7.

**Comment by wsfulton** [comment]
This is unexpected and as there are a multitude of ways to provoke 'undefined behaviour', can you provide a test case that demonstrates the problem? How do you know the size is 9 or 11, are you calling the size() method from Java?

**Comment by mattbru2** [comment]
I am initializing the BoolVector with a size of 10 using new BoolVector(10), and populating it. When I check the size on the C++ side it is not 10.

**Comment by wsfulton** [comment]
Works for me. Try and adapt the example below to show the problem. 
```
~/swig/swig-3.0.7/Examples/scratch/vectorjava$ cat example.i 
%module example

%include <std_vector.i>

%template(BoolVector) std::vector<bool>;

%inline %{
#include <iostream>
void ShowVectorSize(const std::vector<bool>& v) {
  std::cout << "vector size in c++: " << v.size() << std::endl;
}
%}
~/swig/swig-3.0.7/Examples/scratch/vectorjava$ cat runme.java 
public class runme {
  static {
    try {
        System.loadLibrary("example");
    } catch (UnsatisfiedLinkError e) {
      System.err.println("Native code library failed to load. See the chapter on Dynamic Linking Problems in the SWIG Java documentation for help.\n" + e);
      System.exit(1);
    }
  }

  public static void main(String argv[]) {
    BoolVector bv = new BoolVector(10);
    for (int i = 0; i < 10; ++i)
      bv.set(i, true);
    System.out.println("Vector size: " + bv.size());
    example.ShowVectorSize(bv);
  }
}
~/swig/swig-3.0.7/Examples/scratch/vectorjava$ make 
make -f ../../Makefile SRCDIR='' CXXSRCS='example.cxx' SWIG='../../../preinst-swig' \
SWIGOPT='' TARGET='example' INTERFACE='example.i' java_cpp
make[1]: Entering directory '/home/william/swig/swig-3.0.7/Examples/scratch/vectorjava'
../../../preinst-swig -java -c++  -o example_wrap.cxx example.i
g++ -c -fpic -I. -I/usr/include    example.cxx example_wrap.cxx  -I"/usr/lib/jvm/java-1.9.0-openjdk-amd64/include" -I"/usr/lib/jvm/java-1.9.0-openjdk-amd64/include/linux"
g++ -shared  -I/usr/include    example.o example_wrap.o    -o libexample.so
make[1]: Leaving directory '/home/william/swig/swig-3.0.7/Examples/scratch/vectorjava'
make -f ../../Makefile SRCDIR='' JAVASRCS='*.java' JAVAFLAGS='' java_compile
make[1]: Entering directory '/home/william/swig/swig-3.0.7/Examples/scratch/vectorjava'
"javac" -d . *.java
make[1]: Leaving directory '/home/william/swig/swig-3.0.7/Examples/scratch/vectorjava'
make -f ../../Makefile SRCDIR='' java_run
make[1]: Entering directory '/home/william/swig/swig-3.0.7/Examples/scratch/vectorjava'
env LD_LIBRARY_PATH=$PWD  "java" runme 
Vector size: 10
vector size in c++: 10
```

**Comment by mattbru2** [comment]
Thanks for your reply, it helped me solve my issue.

My confusion was in the differences between C++ vector and Java vector/list.
I was using `new BoolVector(10)` followed by 'bv.add(true)',
In Java vector/list, that constructor would create of capacity 10, but size 0, so `add` is a proper function to call.
While in C++ (the BoolVector) that creates a vector of size 10. So my eventual `add` calls were actually making the list larger than 10.

TLDR: I was using `add` instead of `set`
